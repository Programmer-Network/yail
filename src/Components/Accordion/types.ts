import { IDraggableListItem } from "../DraggableList/types";

export interface IAccordionSectionItem {
  id: number;
  title: string;
  order: number;
}

export interface ISection {
  id: number;
  description?: string;
  title: string;
  order: number;
  items: IAccordionSectionItem[];
}

export enum AccordionOrderType {
  SECTIONS = "sections",
  LECTURES = "lectures"
}

export interface IAccordionProps {
  className?: string;
  sections: ISection[];
  sectionTitleClassName?: string;
  onSorted?: (
    type: AccordionOrderType,
    sections: ISection[],
    section: { sectionId: number; items: IDraggableListItem[] }
  ) => void;
  onSectionItemClick?: (item: IDraggableListItem) => void;
  onSectionClick?: (item: IDraggableListItem) => void;
  onSelected?: (item: IDraggableListItem) => void;
  onAddSection?: () => void;
  addSectionItemLabel?: string;
  onAddSectionItem?: (item: ISection) => void;
  addSectionLabel?: string;
  expanded: number[];
  setExpanded: (sections: number[]) => void;
  hasDraggableSections?: boolean;
  hasDraggableSectionItems?: boolean;
  defaultExpanded?: boolean;
  selected?: {
    sectionId: number | null;
    itemId: number | null;
  };
  setSelectedSection?: (sectionId: number | null) => void;
  setSelectedItem?: (itemId: number | null) => void;
}
