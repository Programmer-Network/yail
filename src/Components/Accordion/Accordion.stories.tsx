import { faker } from "@faker-js/faker";
import { action } from "@storybook/addon-actions";
import { Meta } from "@storybook/react";
import { useEffect, useState } from "react";

import Accordion from ".";
import { ISection } from "./types";

const meta = {
  title: "Core / Accordion",
  component: Accordion,
  parameters: {
    layout: "centered"
  },
  tags: ["autodocs"],
  argTypes: {}
} satisfies Meta<typeof Accordion>;

export default meta;

const dummySections = Array(5)
  .fill(0)
  .map((_, index) => ({
    id: Math.floor(Math.random() * 1000000),
    title: faker.lorem.words(Math.floor(Math.random() * 10) + 1),
    description:
      index % 2 === 0
        ? faker.lorem.sentences(Math.floor(Math.random() * 5) + 1)
        : "",

    order: index,
    items: Array(Math.floor(Math.random() * 5) + 1)
      .fill(0)
      .map((__, index) => ({
        id: Math.floor(Math.random() * 1000000),
        title: faker.lorem.words(Math.floor(Math.random() * 10) + 1),
        order: index
      }))
  }));

export const Primary = () => {
  const [sections, setSections] = useState<ISection[]>([]);
  const [expandedSections, setExpandedSections] = useState<number[]>([1]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  useEffect(() => {
    if (sections.length) {
      return;
    }

    setSections(dummySections);
  }, [sections.length]);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        className='yl:w-[500px]'
        sections={sections}
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        onSorted={(_, data) => {
          setSections(data);
        }}
        expanded={expandedSections}
        hasDraggableSections={true}
        hasDraggableSectionItems={true}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
        }}
        onSelected={item => {
          action("onSelected")(item);
        }}
        onSectionItemClick={sectionItem => {
          action("onSectionItemClick")(sectionItem);
        }}
      />
    </div>
  );
};

export const Interactive = () => {
  const [sections, setSections] = useState<ISection[]>([]);
  const [expandedSections, setExpandedSections] = useState<number[]>([1]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  useEffect(() => {
    if (sections.length) {
      return;
    }

    setSections(dummySections);
  }, [sections.length]);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        className='yl:w-[500px]'
        sections={sections}
        expanded={expandedSections}
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
        }}
        onSectionItemClick={sectionItem => {
          action("onSectionItemClick")(sectionItem);
          setSelectedSection(sectionItem.id);
        }}
        onSectionClick={section => {
          action("onSectionClick")(section);
          setSelectedSection(section.id);
        }}
      />
    </div>
  );
};

export const NonInteractiveItems = () => {
  const [sections, setSections] = useState<ISection[]>([]);
  const [expandedSections, setExpandedSections] = useState<number[]>([1]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  useEffect(() => {
    if (sections.length) {
      return;
    }

    setSections(dummySections);
  }, [sections.length]);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        className='yl:w-[500px]'
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        sections={sections}
        onSorted={(_, sections) => {
          setSections(sections);
        }}
        expanded={expandedSections}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
        }}
        onSelected={item => {
          action("onSelected")(item);
        }}
      />
    </div>
  );
};

export const WithoutDragAndDrop = () => {
  const [sections, setSections] = useState<ISection[]>([]);
  const [expandedSections, setExpandedSections] = useState<number[]>([1]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  useEffect(() => {
    if (sections.length) {
      return;
    }

    setSections(dummySections);
  }, [sections.length]);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        className='yl:w-[500px]'
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        sections={sections}
        onSorted={(_, sections) => {
          setSections(sections);
        }}
        expanded={expandedSections}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
        }}
        onSelected={item => {
          action("onSelected")(item);
        }}
      />
    </div>
  );
};

export const WithAddLabels = () => {
  const [sections, setSections] = useState<ISection[]>([]);
  const [expandedSections, setExpandedSections] = useState<number[]>([1]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  useEffect(() => {
    if (sections.length) {
      return;
    }

    setSections(dummySections);
  }, [sections.length]);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        className='yl:w-[500px]'
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        sections={sections}
        expanded={expandedSections}
        onSorted={(_, sections) => {
          setSections(sections);
        }}
        onAddSection={() => {
          action("onAddSection")();
        }}
        addSectionLabel='Add Course Section'
        addSectionItemLabel='Add Course Lecture'
        onAddSectionItem={section => {
          action("onAddSectionItem")(section);
        }}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
          action("setExpanded")(expanded);
        }}
        onSelected={item => {
          action("onSelected")(item);
        }}
      />
    </div>
  );
};

export const WithNoItems = () => {
  const [sections] = useState<ISection[]>([
    {
      id: 1,
      title: "velum constans brevis canis corrumpo magni attollo",
      description: "",
      order: 0,
      items: []
    }
  ]);
  const [expandedSections, setExpandedSections] = useState<number[]>([]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        className='yl:w-[500px]'
        expanded={expandedSections}
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        sections={sections}
        onAddSection={() => {
          action("onAddSection")();
        }}
        addSectionLabel='Add Course Section'
        addSectionItemLabel='Add Course Lecture'
        onAddSectionItem={section => {
          action("onAddSectionItem")(section);
        }}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
          action("setExpanded")(expanded);
        }}
        onSelected={item => {
          action("onSelected")(item);
        }}
      />
    </div>
  );
};

export const ExpandedByDefault = () => {
  const [sections, setSections] = useState<ISection[]>([]);
  const [expandedSections, setExpandedSections] = useState<number[]>([]);
  const [selectedSection, setSelectedSection] = useState<number | null>(null);
  const [selectedItem, setSelectedItem] = useState<number | null>(null);

  useEffect(() => {
    if (sections.length) {
      return;
    }

    setSections(dummySections);
  }, [sections.length]);

  if (!sections.length) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Accordion
        defaultExpanded={true}
        className='yl:w-[500px]'
        selected={{
          sectionId: selectedSection ?? 0,
          itemId: selectedItem ?? 0
        }}
        setSelectedSection={sectionId => {
          setSelectedSection(sectionId);
        }}
        setSelectedItem={itemId => {
          setSelectedItem(itemId);
        }}
        sections={sections}
        expanded={expandedSections}
        onSorted={(_, sections) => {
          setSections(sections);
        }}
        setExpanded={(expanded: number[]) => {
          setExpandedSections(expanded);
          action("setExpanded")(expanded);
        }}
        onSelected={item => {
          action("onSelected")(item);
        }}
      />
    </div>
  );
};
