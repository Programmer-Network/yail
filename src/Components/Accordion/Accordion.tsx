import classNames from "classnames";
import { FC, useEffect, useRef, useState } from "react";

import ArrayUtils from "../../Utils/Array";
import DraggableList from "../DraggableList";
import { IDraggableListItem } from "../DraggableList/types";
import Icon from "../Icon";
import { Paragraph } from "../Typography";
import { AccordionOrderType, IAccordionProps, ISection } from "./types";

const Accordion: FC<IAccordionProps> = ({
  className,
  sections,
  sectionTitleClassName,
  onSectionItemClick,
  onSectionClick,
  onSorted,
  onSelected,
  onAddSection,
  addSectionLabel,
  onAddSectionItem,
  addSectionItemLabel,
  expanded,
  setExpanded,
  hasDraggableSections,
  hasDraggableSectionItems,
  defaultExpanded,
  selected,
  setSelectedSection,
  setSelectedItem
}) => {
  const hasInitialized = useRef(false);
  const [draggedId, setDraggedId] = useState<number | null>(null);
  const [draggedOverId, setDraggedOverId] = useState<number | null>(null);

  const toggleExpand = (sectionId: number) => {
    setExpanded(
      expanded.includes(sectionId)
        ? expanded.filter(id => id !== sectionId)
        : [...expanded, sectionId]
    );
  };

  useEffect(() => {
    if (defaultExpanded && sections.length > 0 && !hasInitialized.current) {
      setExpanded(sections.map(section => section.id));
      hasInitialized.current = true;
    }
  }, [defaultExpanded, sections, setExpanded]);

  const handleDrag = (
    _: React.DragEvent<Element>,
    item: IDraggableListItem
  ) => {
    if (hasDraggableSections) {
      setDraggedId(item.id);
    }
  };

  const handleSectionDrop = (e: React.DragEvent<Element>) => {
    e.preventDefault();
    e.stopPropagation();
    if (draggedId != null && draggedOverId != null) {
      onSorted?.(
        AccordionOrderType.SECTIONS,
        ArrayUtils.reorder(
          sections,
          sections.findIndex(item => item.id === draggedId),
          sections.findIndex(item => item.id === draggedOverId)
        ),
        { sectionId: draggedId, items: [] }
      );
    }

    setDraggedId(null);
    setDraggedOverId(null);
  };

  return (
    <section
      onDragEnd={handleSectionDrop}
      className={classNames("bg-background", className)}
    >
      {sections
        .toSorted((a: ISection, b: ISection) => a.order - b.order)
        .map((section: ISection, idx: number) => (
          <div
            key={section.id}
            draggable={hasDraggableSections && section.items.length > 0}
            onDrag={e => handleDrag(e, section)}
            onDragOver={() =>
              hasDraggableSections && setDraggedOverId(section.id)
            }
            className={classNames(
              "border-border border-2",
              {
                "border-b-0": idx !== sections.length - 1,
                "rounded-tl-md rounded-tr-md": idx === 0,
                "rounded-br-md rounded-bl-md":
                  idx === sections.length - 1 && !onAddSection
              },
              {}
            )}
            role='presentation'
          >
            <h3
              onClick={e => {
                e.stopPropagation();
                e.preventDefault();
                onSectionClick?.(section);
                setSelectedSection?.(section.id);
                setSelectedItem?.(null);
                toggleExpand(section.id);
              }}
              className={classNames(
                "hover:text-primary relative flex cursor-default items-center p-4 font-semibold capitalize transition-colors duration-400 ease-in-out select-none hover:cursor-pointer",
                sectionTitleClassName,
                {
                  "bg-text/2": expanded.includes(section.id),
                  "cursor-pointer": onSectionItemClick,
                  "text-text":
                    selected?.sectionId !== section.id || !onSectionItemClick,
                  "text-primary":
                    onSectionItemClick && selected?.sectionId === section.id
                }
              )}
              onKeyDown={e =>
                ["Enter", " "].includes(e.key) && toggleExpand(section.id)
              }
              tabIndex={0}
              role='button'
              aria-expanded={expanded.includes(section.id)}
            >
              {hasDraggableSections && (
                <div>
                  <Icon
                    iconName='IconDrag'
                    dataTestId='icon-drag'
                    className='mr-1 w-6 opacity-50'
                  />
                </div>
              )}
              <div className='flex flex-col gap-1 overflow-hidden text-base text-ellipsis whitespace-nowrap'>
                {expanded.includes(section.id) ? (
                  <Icon
                    iconName='IconExpandLess'
                    dataTestId='icon-expand-less'
                    className='fill-primary hover:fill-primary absolute right-2 w-6 cursor-pointer'
                  />
                ) : (
                  <Icon
                    iconName='IconExpandMore'
                    dataTestId='icon-expand-more'
                    className='fill-primary hover:fill-primary absolute right-2 w-6 cursor-pointer'
                  />
                )}
                {section.title}
                {section.description && (
                  <Paragraph className='text-text/70 text-xs'>
                    {section.description}
                  </Paragraph>
                )}
              </div>
            </h3>

            {expanded.includes(section.id) && (
              <>
                {section.items.length > 0 && (
                  <DraggableList
                    items={section.items}
                    onClick={(item: IDraggableListItem) => {
                      onSectionItemClick?.(item);
                      onSelected?.(item);
                      setSelectedItem?.(item.id);
                      setSelectedSection?.(
                        sections.find(s => s.items.includes(item))?.id ?? null
                      );
                    }}
                    onDragged={(items: IDraggableListItem[]) => {
                      onSorted?.(
                        AccordionOrderType.LECTURES,
                        [
                          ...sections.map(s =>
                            s.id === section.id ? { ...s, items } : s
                          )
                        ],
                        { sectionId: section.id, items }
                      );
                    }}
                    isDraggable={
                      hasDraggableSectionItems && section.items.length > 0
                    }
                    className='flex flex-col gap-4 py-4'
                    draggedOverClassName='border-t-2 border-border'
                    liClassName={(item: IDraggableListItem) =>
                      classNames(
                        "cursor-default break-words leading-normal px-4",
                        {
                          "hover:cursor-pointer hover:text-primary":
                            onSectionItemClick,
                          "text-primary":
                            onSectionItemClick && selected?.itemId === item.id,
                          "pl-5": section.items.length === 1
                        }
                      )
                    }
                  />
                )}
                {onAddSectionItem && (
                  <div
                    onClick={() => onAddSectionItem(section)}
                    className={classNames(
                      "group border-border hover:bg-text/1 bg-text/3 cursor-pointer"
                    )}
                    role='presentation'
                  >
                    <h3
                      className={classNames(
                        "flex flex-col items-center justify-center p-4 text-center font-semibold select-none"
                      )}
                      role='button'
                    >
                      <Paragraph className='flex items-center gap-1'>
                        <Icon
                          iconName='IconAddCircle'
                          dataTestId='icon-add-circle'
                          className='text-primary w-6'
                        />
                        {addSectionItemLabel && (
                          <span className='text-primary text-sm'>
                            {addSectionItemLabel}
                          </span>
                        )}
                      </Paragraph>
                    </h3>
                  </div>
                )}
              </>
            )}
          </div>
        ))}
      {onAddSection && (
        <div
          onClick={onAddSection}
          className={classNames(
            "group border-border hover:bg-text/2 bg-text/3 cursor-pointer rounded-br-md rounded-bl-md border-2 border-t-0"
          )}
          role='presentation'
        >
          <h3
            className={classNames(
              "flex flex-col items-center justify-center p-4 text-center font-semibold select-none"
            )}
            role='button'
          >
            <Paragraph className='flex items-center gap-1'>
              <Icon
                iconName='IconAddCircle'
                dataTestId='icon-add-circle'
                className='text-primary w-6'
              />
              {addSectionLabel && (
                <span className='text-primary text-sm'>{addSectionLabel}</span>
              )}
            </Paragraph>
          </h3>
        </div>
      )}
    </section>
  );
};

export default Accordion;
