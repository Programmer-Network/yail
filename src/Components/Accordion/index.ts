export { default } from "./Accordion";
export * from "./types";
