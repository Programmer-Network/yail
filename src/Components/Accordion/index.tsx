import classNames from "classnames";
import { FC, useEffect, useState } from "react";

import DraggableList from "Components/DraggableList";
import { IDraggableListItem } from "Components/DraggableList/types";
import {
  IconAddCircle,
  IconDrag,
  IconExpandLess,
  IconExpandMore
} from "Components/Icons";
import { Paragraph } from "Components/Typography";

import ArrayUtils from "Utils/Array";

import { AccordionOrderType, IAccordionProps } from "./types";

const Accordion: FC<IAccordionProps> = ({
  className,
  sections,
  sectionTitleClassName,
  onSectionItemClick,
  onSectionClick,
  onSorted,
  onSelected,
  onAddSection,
  addSectionLabel,
  onAddSectionItem,
  addSectionItemLabel,
  expanded,
  setExpanded,
  selectedId,
  hasDraggableSections,
  hasDraggableSectionItems
}) => {
  const [selectedItemId, setSelectedItemId] = useState<
    number | null | undefined
  >(selectedId);

  const [selectedSectionId, setSelectedSectionId] = useState<
    number | null | undefined
  >(selectedId);

  const [draggedId, setDraggedId] = useState<number | null>(null);
  const [draggedOverId, setDraggedOverId] = useState<number | null>(null);
  const [hoveredId, setHoveredId] = useState<number | null>(null);

  const toggleExpand = (sectionId: number) => {
    setExpanded(
      expanded.includes(sectionId)
        ? expanded.filter(id => id !== sectionId)
        : [...expanded, sectionId]
    );
  };

  useEffect(() => setSelectedItemId(selectedId), [selectedId]);

  const handleDrag = (
    e: React.DragEvent<Element>,
    item: IDraggableListItem
  ) => {
    if (hasDraggableSections) {
      setDraggedId(item.id);
    }
  };

  const handleSectionDrop = (e: React.DragEvent<Element>) => {
    e.preventDefault();
    e.stopPropagation();
    if (draggedId != null && draggedOverId != null) {
      onSorted?.(
        AccordionOrderType.SECTIONS,
        ArrayUtils.reorder(
          sections,
          sections.findIndex(item => item.id === draggedId),
          sections.findIndex(item => item.id === draggedOverId)
        ),
        { sectionId: draggedId, items: [] }
      );
    }

    setDraggedId(null);
    setDraggedOverId(null);
  };

  return (
    <section onDragEnd={handleSectionDrop} className={classNames(className)}>
      {sections
        .toSorted((a, b) => a.order - b.order)
        .map((section, idx) => (
          <div
            key={section.id}
            draggable={hasDraggableSections}
            onDrag={e => handleDrag(e, section)}
            onDragOver={() =>
              hasDraggableSections && setDraggedOverId(section.id)
            }
            onMouseOver={() => hasDraggableSections && setHoveredId(section.id)}
            onMouseLeave={() => hasDraggableSections && setHoveredId(null)}
            className={classNames(
              "yl-border-2 yl-border-border",
              {
                "yl-border-b-0": idx !== sections.length - 1,
                "yl-rounded-tl-md yl-rounded-tr-md": idx === 0,
                "yl-rounded-bl-md yl-rounded-br-md":
                  idx === sections.length - 1 && !onAddSection
              },
              {
                "yl-bg-text/5": draggedOverId === section.id,
                "yl-bg-text/10": draggedId && hoveredId === section.id
              }
            )}
            role='presentation'
          >
            <h3
              onClick={e => {
                e.stopPropagation();
                e.preventDefault();
                onSectionClick?.(section);
                setSelectedSectionId(section.id);
              }}
              className={classNames(
                "yl-relative yl-flex yl-cursor-default yl-select-none yl-items-center yl-font-semibold yl-capitalize yl-p-4",
                sectionTitleClassName,
                {
                  "yl-bg-text/5": expanded.includes(section.id),
                  "yl-cursor-pointer": onSectionItemClick,
                  "yl-text-text":
                    selectedSectionId !== section.id || !onSectionItemClick,
                  "yl-text-primary":
                    onSectionItemClick && selectedSectionId === section.id
                }
              )}
              onKeyDown={e =>
                ["Enter", " "].includes(e.key) && toggleExpand(section.id)
              }
              tabIndex={0}
              role='button'
              aria-expanded={expanded.includes(section.id)}
            >
              {hasDraggableSections && (
                <div>
                  <IconDrag className='yl-w-6 yl-opacity-50 yl-mr-1' />
                </div>
              )}
              <div className='yl-flex yl-flex-col yl-gap-1 yl-overflow-hidden yl-text-ellipsis yl-whitespace-nowrap yl-text-base'>
                {expanded.includes(section.id) ? (
                  <IconExpandLess
                    onClick={e => {
                      e.stopPropagation();
                      e.preventDefault();
                      toggleExpand(section.id);
                    }}
                    className='yl-absolute yl-right-2 yl-w-6 yl-cursor-pointer yl-fill-primary hover:yl-fill-primary'
                  />
                ) : (
                  <IconExpandMore
                    onClick={e => {
                      e.stopPropagation();
                      e.preventDefault();
                      toggleExpand(section.id);
                    }}
                    className='yl-absolute yl-right-2 yl-w-6 yl-cursor-pointer yl-fill-primary hover:yl-fill-primary'
                  />
                )}
                {section.title}
                {section.description && (
                  <Paragraph className='yl-text-xs yl-text-text/70'>
                    {section.description}
                  </Paragraph>
                )}
              </div>
            </h3>

            {expanded.includes(section.id) && (
              <>
                <DraggableList
                  items={section.items}
                  onClick={(item: IDraggableListItem) => {
                    onSectionItemClick?.(item);
                    onSelected?.(item);
                    setSelectedItemId(item.id);
                  }}
                  onDragged={(items: IDraggableListItem[]) => {
                    onSorted?.(
                      AccordionOrderType.LECTURES,
                      [
                        ...sections.map(s =>
                          s.id === section.id ? { ...s, items } : s
                        )
                      ],
                      { sectionId: section.id, items }
                    );
                  }}
                  isDraggable={hasDraggableSectionItems}
                  className='yl-gap-4 yl-flex yl-flex-col yl-py-4'
                  draggedOverClassName='yl-border-t-2 yl-border-border'
                  liClassName={(item: IDraggableListItem) =>
                    classNames(
                      "yl-cursor-default hover:text-text break-words leading-normal yl-px-4",
                      {
                        "yl-cursor-pointer": onSectionItemClick,
                        "yl-text-text": selectedItemId !== item.id,
                        // "yl-text-text":
                        //   onSectionItemClick && selectedItemId === item.id,
                        "yl-pl-5": section.items.length === 1
                      }
                    )
                  }
                />
                {onAddSectionItem && (
                  <div
                    onClick={() => onAddSectionItem(section)}
                    className={classNames(
                      "yl-cursor-pointer hover:yl-text-text yl-pb-4 yl-text-text yl-text-center yl-flex yl-items-center yl-justify-center yl-gap-1"
                    )}
                  >
                    <IconAddCircle className='yl-w-6' />
                    {addSectionItemLabel && (
                      <span className='yl-text-sm'>{addSectionItemLabel}</span>
                    )}
                  </div>
                )}
              </>
            )}
          </div>
        ))}
      {onAddSection && (
        <div
          onClick={onAddSection}
          className={classNames(
            "yl-group yl-border-2 yl-border-t-0 yl-border-border hover:yl-bg-text/5 yl-rounded-bl-md yl-rounded-br-md yl-cursor-pointer"
          )}
          role='presentation'
        >
          <h3
            className={classNames(
              "yl-flex yl-flex-col yl-items-center yl-justify-center yl-select-none yl-font-semibold yl-p-4 yl-text-center"
            )}
            role='button'
          >
            <Paragraph className='yl-text-text/70 group-hover:yl-text-text yl-flex yl-items-center yl-gap-1'>
              <IconAddCircle className='yl-w-6 group-hover:yl-text-text' />
              {addSectionLabel && (
                <span className='yl-text-sm'>{addSectionLabel}</span>
              )}
            </Paragraph>
          </h3>
        </div>
      )}
    </section>
  );
};

export default Accordion;
