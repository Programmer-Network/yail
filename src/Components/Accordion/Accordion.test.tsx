import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import Accordion from ".";

const mockSections = [
  {
    id: 1,
    description: "Description 1",
    title: "Section 1",
    order: 1,
    items: [
      { id: 1, title: "Item A", order: 1 },
      { id: 2, title: "Item B", order: 2 }
    ]
  },
  {
    id: 2,
    description: "Description 2",
    title: "Section 2",
    order: 2,
    items: [{ id: 3, title: "Item C", order: 3 }]
  }
];

describe("Accordion component", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <Accordion
        sections={mockSections}
        onSorted={vi.fn()}
        setExpanded={() => null}
        expanded={[1]}
        onSectionItemClick={vi.fn()}
        selected={{ sectionId: 0, itemId: 0 }}
        setSelectedSection={() => null}
        setSelectedItem={() => null}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders all sections and items correctly", () => {
    const setExpanded = vi.fn();
    render(
      <Accordion
        sections={mockSections}
        onSorted={vi.fn()}
        setExpanded={setExpanded}
        expanded={[1]}
        onSectionItemClick={vi.fn()}
        selected={{ sectionId: 0, itemId: 0 }}
        setSelectedSection={() => null}
        setSelectedItem={() => null}
      />
    );

    expect(screen.getByText("Section 1")).toBeInTheDocument();
    expect(screen.getByText("Item A")).toBeInTheDocument();
    expect(screen.getByText("Section 2")).toBeInTheDocument();
    expect(screen.queryByText("Item C")).toBeNull();
  });

  test("expands and collapses sections on click", () => {
    const setExpanded = vi.fn();
    const setSelectedSection = vi.fn();
    const setSelectedItem = vi.fn();

    render(
      <Accordion
        sections={mockSections}
        onSorted={vi.fn()}
        setExpanded={setExpanded}
        expanded={[mockSections[0]!.id]}
        onSectionItemClick={vi.fn()}
        selected={{ sectionId: mockSections[0]!.id, itemId: 0 }}
        setSelectedSection={setSelectedSection}
        setSelectedItem={setSelectedItem}
      />
    );

    fireEvent.click(screen.getByTestId("icon-expand-less"));
    expect(setExpanded).toHaveBeenCalledWith([]);
  });

  test("handles section item click events", () => {
    const onSectionItemClick = vi.fn();
    const setSelectedSection = vi.fn();
    const setSelectedItem = vi.fn();

    render(
      <Accordion
        sections={mockSections}
        onSorted={vi.fn()}
        setExpanded={vi.fn()}
        onSectionItemClick={onSectionItemClick}
        expanded={[1]}
        selected={{ sectionId: 1, itemId: 0 }}
        setSelectedSection={setSelectedSection}
        setSelectedItem={setSelectedItem}
      />
    );

    fireEvent.click(screen.getByText("Item A"));
    expect(onSectionItemClick).toHaveBeenCalledWith(mockSections[0]!.items[0]!);
    expect(setSelectedItem).toHaveBeenCalledWith(mockSections[0]!.items[0]!.id);
    expect(setSelectedSection).toHaveBeenCalledWith(mockSections[0]!.id);
  });

  test("applies custom class names", () => {
    render(
      <Accordion
        sections={mockSections}
        onSorted={vi.fn()}
        className='custom-class'
        sectionTitleClassName='custom-title-class'
        setExpanded={vi.fn()}
        expanded={[]}
        onSectionItemClick={vi.fn()}
        selected={{ sectionId: 0, itemId: 0 }}
        setSelectedSection={() => null}
        setSelectedItem={() => null}
      />
    );

    expect(screen.getByText("Section 1").parentNode).toHaveClass(
      "custom-title-class"
    );
  });

  test("expands all sections by default when defaultExpanded is true", async () => {
    const setExpanded = vi.fn();
    render(
      <Accordion
        sections={mockSections}
        onSorted={vi.fn()}
        setExpanded={setExpanded}
        expanded={[]}
        defaultExpanded={true}
        onSectionItemClick={vi.fn()}
        selected={{ sectionId: 0, itemId: 0 }}
        setSelectedSection={() => null}
        setSelectedItem={() => null}
      />
    );

    expect(setExpanded).toHaveBeenCalledWith(
      mockSections.map(section => section.id)
    );

    mockSections.forEach(section => {
      expect(screen.getByText(section.title)).toBeInTheDocument();
    });
  });
});
