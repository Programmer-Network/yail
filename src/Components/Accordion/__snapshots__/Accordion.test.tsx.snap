// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion component > renders correctly 1`] = `
<DocumentFragment>
  <section
    class=""
  >
    <div
      class="yl:border-2 yl:border-border yl:border-b-0 yl:rounded-tl-md yl:rounded-tr-md"
      role="presentation"
    >
      <h3
        aria-expanded="true"
        class="yl:relative yl:flex yl:cursor-default yl:select-none yl:items-center yl:font-semibold yl:capitalize yl:p-4 yl:bg-text/5 yl:cursor-pointer yl:text-text"
        role="button"
        tabindex="0"
      >
        <div
          class="yl:flex yl:flex-col yl:gap-1 yl:overflow-hidden yl:text-ellipsis yl:whitespace-nowrap yl:text-base"
        >
          <div
            class="yl:absolute yl:right-2 yl:w-6 yl:cursor-pointer yl:fill-primary yl:hover:fill-primary"
            data-icon="IconExpandLess"
            data-testid="icon-expand-less"
          >
            [Icon: IconExpandLess]
          </div>
          Section 1
          <p
            class="yl:text-text yl:break-words yl:overflow-hidden yl:text-xs yl:text-text/70"
          >
            Description 1
          </p>
        </div>
      </h3>
      <ul
        class="yl:relative yl:text-text yl:gap-4 yl:flex yl:flex-col yl:py-4"
      >
        <li
          class="yl:relative yl:flex yl:items-center yl:cursor-default hover:text-text break-words leading-normal yl:px-4 yl:cursor-pointer yl:text-text"
        >
          Item A
        </li>
        <li
          class="yl:relative yl:flex yl:items-center yl:cursor-default hover:text-text break-words leading-normal yl:px-4 yl:cursor-pointer yl:text-text"
        >
          Item B
        </li>
      </ul>
    </div>
    <div
      class="yl:border-2 yl:border-border yl:rounded-bl-md yl:rounded-br-md"
      role="presentation"
    >
      <h3
        aria-expanded="false"
        class="yl:relative yl:flex yl:cursor-default yl:select-none yl:items-center yl:font-semibold yl:capitalize yl:p-4 yl:cursor-pointer yl:text-text"
        role="button"
        tabindex="0"
      >
        <div
          class="yl:flex yl:flex-col yl:gap-1 yl:overflow-hidden yl:text-ellipsis yl:whitespace-nowrap yl:text-base"
        >
          <div
            class="yl:absolute yl:right-2 yl:w-6 yl:cursor-pointer yl:fill-primary yl:hover:fill-primary"
            data-icon="IconExpandMore"
            data-testid="icon-expand-more"
          >
            [Icon: IconExpandMore]
          </div>
          Section 2
          <p
            class="yl:text-text yl:break-words yl:overflow-hidden yl:text-xs yl:text-text/70"
          >
            Description 2
          </p>
        </div>
      </h3>
    </div>
  </section>
</DocumentFragment>
`;
